src/main/java/org/folio/service/PasswordResetLinkServiceImpl.java

```diff
@@ -0,0 +1,76 @@
+import org.folio.rest.util.OkapiConnectionParams;
+
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Collections;
+
+public class PasswordResetLinkServiceImpl implements PasswordResetLinkService {
+
+  private AuthTokenClient authTokenClient;
+  private PasswordResetActionClient passwordResetActionClient;
+  private UserModuleClient userModuleClient;
+
+  public PasswordResetLinkServiceImpl(AuthTokenClient authTokenClient,
+                                      PasswordResetActionClient passwordResetActionClient,
+                                      UserModuleClient userModuleClient) {
+    this.authTokenClient = authTokenClient;
+    this.passwordResetActionClient = passwordResetActionClient;
+    this.userModuleClient = userModuleClient;
+  }
+
+  @Override
+  public Future<TokenResponse> validateLinkAndLoginUser(OkapiConnectionParams okapiConnectionParams) {
+    String token = okapiConnectionParams.getToken();
+    JsonObject payload = new JsonObject(Buffer.buffer(Base64.getDecoder().decode(token.split("\\.")[1])));
+    String passwordResetActionId = payload.getString("passwordResetActionId");
+
+    return passwordResetActionClient.getAction(passwordResetActionId, okapiConnectionParams)
+      .compose(pwdResetAction -> {
+        if (pwdResetAction.isPresent()) {
+          return Future.succeededFuture(pwdResetAction.get());
+        } else {
+          UnprocessableEntityMessage message = new UnprocessableEntityMessage("action.not.found",
+            String.format("PasswordResetAction with id = %s is not found", passwordResetActionId));
+          return Future.failedFuture(new UnprocessableEntityException(Collections.singletonList(message)));
+        }
+      }).compose(pwdResetAction -> {
+        if (pwdResetAction.getExpirationTime().toInstant().isAfter(Instant.now())) {
+          return Future.succeededFuture(pwdResetAction);
+        } else {
+          UnprocessableEntityMessage message = new UnprocessableEntityMessage("action.expired",
+            String.format("PasswordResetAction with id = %s is expired", passwordResetActionId));
+          return Future.failedFuture(new UnprocessableEntityException(Collections.singletonList(message)));
+        }
+      }).compose(pwdResetAction -> userModuleClient.lookupUserById(pwdResetAction.getUserId(), okapiConnectionParams)
+        .map(user -> {
+        if (user.isPresent()) {
+          return Future.succeededFuture(user.get());
+        } else {
+          UnprocessableEntityMessage message = new UnprocessableEntityMessage("user.not.found",
+            String.format("User with id = %s in not found", pwdResetAction.getUserId()));
+          return Future.<User>failedFuture(new UnprocessableEntityException(Collections.singletonList(message)));
+        }
+      })).compose(user -> {
+        JsonObject tokenPayload = new JsonObject().put("payload",
+          new JsonObject().put("sub", user.result().getUsername()).put("user_id", user.result().getId()));
```

Comment:
authTokenClient puts given JsonObject to property "payload", so there is no need to do it here.